### PR TITLE
APP-1088: External insurer view

### PR DESCRIPTION
**Pull request by hugokallstrom**
### Checklist

- [ ] Functionality is covered by an integration test
- [ ] Functionality is accessible in engineering mode

### CODE REVIEW

File: app/src/main/java/com/hedvig/app/feature/embark/passages/externalinsurer/ExternalInsurerFragment.kt [hunk 1]
```diff
@@ -3,30 +3,91 @@ package com.hedvig.app.feature.embark.passages.externalinsurer
...
 
     private val binding by viewBinding(PreviousOrExternalInsurerFragmentBinding::bind)
+
+    private val embarkViewModel: EmbarkViewModel by sharedViewModel()
     private val viewModel: ExternalInsurerViewModel by sharedViewModel()
 
+    private val askForPriceActivityResultLauncher =
+        registerForActivityResult(ActivityResultContracts.StartActivityForResult()) { result: ActivityResult ->
+            if (result.resultCode == RESULT_SKIP) {
+                onContinue()
```
Comment (hugokallstrom) [inline app/src/main/java/com/hedvig/app/feature/embark/passages/externalinsurer/ExternalInsurerFragment.kt:42]: I'm a bit unsure how we handle "switching" here, that was previously done with `PreviousInsurer`. @sampettersson  how is this implemented on iOS? Eg. A users selects "skip" in insurely, but has still selected their previous insurer. Are we showing switching information in the offer screen?

File: app/src/main/java/com/hedvig/app/feature/embark/passages/externalinsurer/ExternalInsurerFragment.kt [hunk 2]
```diff
@@ -44,19 +105,53 @@ class ExternalInsurerFragment : Fragment(R.layout.previous_or_external_insurer_f
...
+            }
+            continueButton.setOnClickListener {
+                viewModel.onContinue()
+            }
+
+            setFragmentResultListener(InsurerProviderBottomSheet.REQUEST_KEY) { requestKey: String, bundle: Bundle ->
+                val id = bundle.getString(InsurerProviderBottomSheet.INSURER_ID_KEY)
+                val name = bundle.getString(InsurerProviderBottomSheet.INSURER_NAME_KEY)
+                if (requestKey == InsurerProviderBottomSheet.REQUEST_KEY && id != null && name != null) {
+                    // embarkViewModel.putInStore(insurerData.storeKey, item.id)
```
Comment (hugokallstrom) [inline app/src/main/java/com/hedvig/app/feature/embark/passages/externalinsurer/ExternalInsurerFragment.kt:120]: This is also related to the question regarding switching

File: app/src/main/java/com/hedvig/app/feature/embark/passages/externalinsurer/ExternalInsurerViewModel.kt [hunk 3]
```diff
@@ -2,27 +2,39 @@ package com.hedvig.app.feature.embark.passages.externalinsurer
...
 class ExternalInsurerViewModel(
     private val getInsuranceProvidersUseCase: GetInsuranceProvidersUseCase
 ) : ViewModel() {
 
+    private var insuranceProviders: List<InsuranceProvider> = emptyList()
+
+    private val _events = Channel<Event>(Channel.UNLIMITED)
+    val events = _events.receiveAsFlow()
+
+    sealed class Event {
```
Comment (hugokallstrom) [inline app/src/main/java/com/hedvig/app/feature/embark/passages/externalinsurer/ExternalInsurerViewModel.kt:22]: @hedvigoscar I moved this to an event here

File: app/src/main/java/com/hedvig/app/feature/embark/passages/externalinsurer/ExternalInsurerFragment.kt (hunk 1 above)
Comment (sampettersson) [reply app/src/main/java/com/hedvig/app/feature/embark/passages/externalinsurer/ExternalInsurerFragment.kt:42]: Yes, you should still set the previous insurer in the store (don't set the dataCollectionId though), regard skipping as just saying "I don't want to use insurely, but I still want to switch".

File: app/src/main/java/com/hedvig/app/feature/embark/passages/externalinsurer/ExternalInsurerFragment.kt [hunk 4]
```diff
@@ -3,30 +3,91 @@ package com.hedvig.app.feature.embark.passages.externalinsurer
...
+        viewModel.viewState
+            .flowWithLifecycle(lifecycle)
+            .onEach { viewState ->
+                binding.progress.isVisible = viewState.isLoading
+
+                viewState.error?.let {
+                    context?.showErrorDialog(getString(it.getStringRes())) { }
+                }
+
+                viewState.showInsuranceProviders?.let {
```
Comment (hedvigoscar) [inline app/src/main/java/com/hedvig/app/feature/embark/passages/externalinsurer/ExternalInsurerFragment.kt:80]: What happens on configuration change here? `onShowInsurers` launches a new bottom sheet, wouldn't we get two bottom sheets in that case?

Seems like this should perhaps be an event 🤔

File: app/src/main/java/com/hedvig/app/feature/embark/passages/externalinsurer/ExternalInsurerViewModel.kt [hunk 5]
```diff
@@ -2,27 +2,39 @@ package com.hedvig.app.feature.embark.passages.externalinsurer
...
+import kotlinx.coroutines.flow.asStateFlow
+import kotlinx.coroutines.flow.receiveAsFlow
 import kotlinx.coroutines.flow.update
 import kotlinx.coroutines.launch
 
 class ExternalInsurerViewModel(
     private val getInsuranceProvidersUseCase: GetInsuranceProvidersUseCase
 ) : ViewModel() {
 
+    private var insuranceProviders: List<InsuranceProvider> = emptyList()
```
Comment (hedvigoscar) [inline app/src/main/java/com/hedvig/app/feature/embark/passages/externalinsurer/ExternalInsurerViewModel.kt:17]: Is there a way to model this where we're not using `var` here? I'm always a bit wary of seeing `var` 😧

Comment (hedvigoscar) [reply app/src/main/java/com/hedvig/app/feature/embark/passages/externalinsurer/ExternalInsurerViewModel.kt:17]: Perhaps the `insuranceProviders` could always be a part of the `ViewState`, and then the interaction of launching the bottom sheet would never require a round trip to the `ViewModel` 🤔

File: app/src/main/java/com/hedvig/app/feature/embark/passages/previousinsurer/PreviousInsurerFragment.kt [hunk 6]
```diff
@@ -1,42 +1,29 @@
...
 import com.hedvig.app.util.extensions.view.setHapticClickListener
 import com.hedvig.app.util.extensions.view.setupInsetsForIme
-import com.hedvig.app.util.featureflags.Feature
-import com.hedvig.app.util.featureflags.FeatureManager
 import com.hedvig.app.util.whenApiVersion
 import com.zhuinden.fragmentviewbindingdelegatekt.viewBinding
-import org.koin.android.ext.android.inject
+import d
 import org.koin.androidx.viewmodel.ext.android.sharedViewModel
+import timber.log.Timber.Forest.d
```
Comment (hedvigoscar) [inline app/src/main/java/com/hedvig/app/feature/embark/passages/previousinsurer/PreviousInsurerFragment.kt:20]: What's this? 🤔

File: app/src/main/java/com/hedvig/app/feature/settings/Market.kt [hunk 7]
```diff
@@ -75,8 +74,7 @@ enum class Market {
     }
 
     fun onboarding(context: Context) = when (this) {
-        SE -> ChatActivity.newInstance(context)
-            .apply { putExtra(ChatActivity.EXTRA_SHOW_RESTART, true) }
```
Comment (hedvigoscar) [inline app/src/main/java/com/hedvig/app/feature/settings/Market.kt:79]: Should this be an argument to `ChatActivity.Companion::newInstance`?

Comment (hedvigoscar) [reply app/src/main/java/com/hedvig/app/feature/settings/Market.kt:79]: Nevermind, this was just removed code. My bad!

**Pull request merged**